### PR TITLE
docker:login and docker:push with the new docker runtime/registry

### PR DESCRIPTION
**Pull request by hunterloftis**
Adds two commands for our new docker runtime and registry:

- `heroku docker:login`
- `heroku docker:push`

To test:

- `heroku sudo labs:enable docker-releases` (flag yourself into the docker runtime)
- clone, checkout this branch
- from within this repo's dir, `heroku plugins:link` (tell the cli to symlink a local plugin)
- cd into a directory with a (not important) Dockerfile, an app, and a git repo
- `git remote rm heroku` (disconnect any existing heroku app)
- `heroku create`
- `heroku docker:login`
- `heroku docker:push`
- `heroku open`

Notes and details:

- https://gist.github.com/tt/b81be8de690441a791bdfe507f947954

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
   "license": "ISC",
   "dependencies": {
     "camelcase": "1.0.2",
+    "co": "^4.6.0",
```
Comment (dickeyxxx) [inline package.json]: usually good not to let packages autoupdate, we've had problems in the past with this

**Comment by rsampaio** [comment]
I'm wondering if generate or manipulate .docker/config.json file would be a better option for the auth part, the format is pretty simple and would require one less docker cli shell out: 

```json
{
	"auths": {
		"localhost:8080": {
			"auth": "<auth base64 encoded>",
			"email": "rodrigo@heroku.com"
		}
	}
}
```

File: package.json (hunk 1 above)
Comment (hunterloftis) [inline package.json]: I agree; initially I updated and fixed all the deps, but I'd rather do that as a totally separate PR, so I just went with the flow here. Want to make sure I can get :login and :push in without breaking any existing docker users, so minimal changes.

**Comment by hunterloftis** [comment]
@rsampaio I'm following this pattern at the moment: https://gist.github.com/tt/b81be8de690441a791bdfe507f947954

The additional docker shell could prove to be a benefit - preventing people from going through the whole push and then getting an auth failure (@tt is looking into validating that you're flagged into docker on :login, which currently doesn't happen).

I'll leave it to you docker experts to let me know which way you prefer for auth.

**Comment by hunterloftis** [comment]
Another option that wouldn't require any config file writing and that would also skip an extra command is basically running the `docker login` automatically before `docker push`. Login is pretty quick and that would help to get a fast failure (auth failures happen at the end of :push right now). /cc @tt

**Comment by tt** [comment]
I prefer not writing Docker's configuration file. It's format or location has already changed once and if we write it, it's our responsibility to not set to wide permissions etc. Shelling out also validates the provided credentials (and allows us to provide feedback due to the round trip).

When we launch this, I think `heroku login` should do the right thing. If anyone feels strongly that we should pursue the workaround where we shell out before executing `docker push`, that works for me but it's some extra requests per invocation.

**Comment by hunterloftis** [comment]
@tt re: `heroku login`, do you mean this plugin should hook into `heroku login` and automatically `docker login` as well? We can open a separate issue for that, but I suspect it may be a little trickier. For instance, I'm imagining it on my osx laptop, where I have to do a series of docker-machine invocations before `docker` is even available as a command.

File: commands/push.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    command: 'push',
+    description: 'Deploys a Docker image to your Heroku app',
+    needsApp: true,
+    needsAuth: true,
+    args: [{ name: 'process', optional: true }],
+    run: cli.command(co.wrap(release))
+  };
+};
+
+function* release(context, heroku) {
```
Comment (tt) [inline commands/push.js]: `push`?

File: commands/push.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  let herokuHost = process.env.HEROKU_HOST || 'heroku.com';
+  let registry = `registry.${ herokuHost }`;
+  let password = context.auth.password;
```
Comment (tt) [inline commands/push.js]: This is unused.

File: commands/push.js (hunk 2 above)
Comment (hunterloftis) [inline commands/push.js]: Ha good catch. I'll also add to the help that is builds and then pushes the image to your app, for clarity.

File: commands/push.js (hunk 3 above)
Comment (hunterloftis) [inline commands/push.js]: Do you mean I don't need to include `--password` with the login request?

Comment (hunterloftis) [inline commands/push.js]: Oh, I see, this is on push. copypasta, thanks.

**Comment by tt** [comment]
Looks good to me.

**Pull request merged**